src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -26,520 +31,240 @@
-    public abstract Object appendArray(Object store, int index, DynamicObject array);
-    public abstract Object appendValue(Object store, int index, Object value);
-    public abstract Object finish(Object store, int length);
-
-    protected Object restart(int length, String reason) {
-        final UninitializedArrayBuilderNode newNode = new UninitializedArrayBuilderNode();
-        replace(newNode, reason);
-        return newNode.start(length);
-    }
-
-    protected Object appendValueFallback(Object store, int index, Object value, int expectedLength) {
-        replace(new ObjectArrayBuilderNode(expectedLength));
-
-        // The store type cannot be assumed if multiple threads use the same builder,
-        // so just use the generic box() since anyway this is slow path.
-        final Object[] newStore;
-        if (store.getClass() == Object[].class) {
-            newStore = (Object[]) store;
-        } else {
-            newStore = ArrayUtils.box(store);
-        }
-
-        newStore[index] = value;
-        return newStore;
-    }
-
-    protected Object ensureFallback(Object store, int length) {
-        final Object[] newStore = ArrayUtils.box(store, length);
-        final UninitializedArrayBuilderNode newNode = new UninitializedArrayBuilderNode();
-        replace(newNode);
-        newNode.resume(newStore);
-        return newNode.ensure(newStore, length);
-    }
-
-    private static class UninitializedArrayBuilderNode extends ArrayBuilderNode {
-
-        private boolean couldUseInteger = true;
-        private boolean couldUseLong = true;
-        private boolean couldUseDouble = true;
-        private final BranchProfile expandProfile = BranchProfile.create();
-
-        public void resume(Object[] store) {
-            for (Object value : store) {
-                screen(value);
-            }
-        }
-
-        @Override
-        public Object start() {
-            return new Object[getContext().getOptions().ARRAY_UNINITIALIZED_SIZE];
-        }
-
-        @Override
-        public Object start(int length) {
-            return new Object[length];
-        }
-
-        @Override
-        public Object ensure(Object store, int length) {
-            // All appends go through append(Object, int, Object), which is always happy to make space
             return store;
         }
 
-        @TruffleBoundary
-        @Override
-        public Object appendArray(Object store, int index, DynamicObject array) {
-            for (Object value : ArrayOperations.toIterable(array)) {
-                store = appendValue(store, index, value);
-                index++;
+        public void updateStrategy(ArrayStrategy strategy, int size) {
+            if (startNode.replacement(strategy, size)) {
+                ensureNode.replace(EnsureCapacityNode.create());
+                appendArrayNode.replace(AppendArrayNode.create());
+                appendOneNode.replace(AppendOneNode.create());
```

Comment:
Since those nodes seem to assume a single fixed strategy, should they just take an ArrayStrategy in their constructor then?
Is there a guarantee one strategy is always more general than others?
I guess currently this is true for ObjectArrayStrategy, and since this node only creates new arrays, concurrent strategies should not be involved.

Reply:
> Since those nodes seem to assume a single fixed strategy, should they just take an ArrayStrategy in their constructor then?

@aardvark179 What do you think?